src/Microsoft.ML.Data/Utilities/LocalEnvironment.cs

```diff
@@ -87,6 +92,13 @@ protected override IPipe<TMessage> CreatePipe<TMessage>(ChannelProviderBase pare
             return new Pipe<TMessage>(parent, name, GetDispatchDelegate<TMessage>());
         }
 
+        public override CompositionContainer GetCompositionContainer()
+        {
+            if (_compositionContainerFactory != null)
+                return _compositionContainerFactory();
```

Comment:
Do we really want to make a new `CompositionContainer` every time it is asked for? If so, I think it would make more sense to call this `CreateCompositionContainer()`, to inform the caller that this is a new one every time.

Reply:
I don't see any other way, since the user may be adding more components to `PartCatalog` over time

---
In reply to: [229417995](https://github.com/dotnet/machinelearning/pull/1406#discussion_r229417995) [](ancestors = 229417995)